### PR TITLE
Make popup inert when destroyed.

### LINKED ISSUE #1417: Remove wlr_xdg_surface_send_close
Replace with two separate functions: `wlr_xdg_toplevel_send_close` and `wlr_xdg_popup_destroy`.

The problem is that closing a popup has different semantics from closing a toplevel.

- When closing a toplevel, the compositor merely asks the client to shut down -- the client can ignore the request
- When closing a popup, we notify the client that the popup has been closed, the client cannot do anything about it. In this case we can destroy (aka `free`) the `wlr_xdg_surface` and make the resource inert.

### PR DESCRIPTION
**Pull request by c-edw**
Closes #1417.

### CODE REVIEW

**Comment by c-edw** [comment]
Sorry for the delay, I haven't been feeling great. Are you asking me to swap out the asserts for NULL checks?

**Comment by emersion** [comment]
>Sorry for the delay, I haven't been feeling great.

No worries. Take your time, we're not in a hurry. Take some time off from contributing if you need to. :)

>Are you asking me to swap out the asserts for NULL checks?

Maybe changed to something like this?

```c
assert(xdg_surface->role == WLR_XDG_SURFACE_ROLE_NONE || xdg_surface->role == WLR_XDG_SURFACE_ROLE_POPUP);
```

(Since `reset_xdg_surface` will set the role to "none")

**Comment by c-edw** [comment]
Thanks, I appreciate it. I have read the section in CONTRIBUTING.md about inert resources like I probably should have. It says that handlers should ignore inert resources, but I'm not sure if this also means other code should pretend that these inert resources don't exist. I assumed yes.

File: types/xdg_shell/wlr_xdg_surface.c [hunk 1]
```diff
@@ -524,13 +524,22 @@ void wlr_xdg_toplevel_send_close(struct wlr_xdg_surface *surface) {
...
+	if (surface == NULL) {
+		return;
+	}
 	assert(surface->popup);
-	assert(surface->role == WLR_XDG_SURFACE_ROLE_POPUP);
+	assert(surface->role == WLR_XDG_SURFACE_ROLE_POPUP ||
+		surface->role == WLR_XDG_SURFACE_ROLE_NONE);
 	xdg_popup_send_popup_done(surface->popup->resource);
+	wl_resource_set_user_data(surface->popup->resource, NULL);
+	destroy_xdg_popup(surface->popup->base);
```
Comment (emersion) [inline types/xdg_shell/wlr_xdg_surface.c]: I believe this can just be `surface`.

File: types/xdg_shell/wlr_xdg_surface.c [hunk 2]
```diff
@@ -546,6 +555,9 @@ struct wlr_surface *wlr_xdg_surface_surface_at(
 		double *sub_x, double *sub_y) {
 	struct wlr_xdg_popup *popup_state;
 	wl_list_for_each(popup_state, &surface->popups, link) {
+		if (popup_state == NULL) {
```
Comment (emersion) [inline types/xdg_shell/wlr_xdg_surface.c]: This isn't possible. `destroy_xdg_popup` will remove the popup from the list.

File: types/xdg_shell/wlr_xdg_surface.c [hunk 3]
```diff
@@ -524,13 +524,22 @@ void wlr_xdg_toplevel_send_close(struct wlr_xdg_surface *surface) {
...
+	assert(surface->role == WLR_XDG_SURFACE_ROLE_POPUP ||
+		surface->role == WLR_XDG_SURFACE_ROLE_NONE);
 	xdg_popup_send_popup_done(surface->popup->resource);
+	wl_resource_set_user_data(surface->popup->resource, NULL);
+	destroy_xdg_popup(surface->popup->base);
 }
 
 static void xdg_popup_get_position(struct wlr_xdg_popup *popup,
 		double *popup_sx, double *popup_sy) {
+	if (popup == NULL) {
```
Comment (emersion) [inline types/xdg_shell/wlr_xdg_surface.c]: NULL checks should be added in protocol request handlers, not here. That means `xdg_popup_handle_grab`, but apparently that's already done, so we're good.

**Comment by emersion** [comment]
Thanks!

**Pull request merged**